py3status/modules/keyboard_layout.py

```diff
@@ -3,9 +3,15 @@
 Display keyboard layout.
 
 Configuration parameters:
+    button_next: Button to click to cycle up through available layouts
+        (default 4)
+    button_prev: Button to click to cycle up through available layouts
+        (default 5)
+    button_select: Button to click to select current layout (default 1)
     cache_timeout: refresh interval for this module (default 10)
     colors: deprecated. see color options below (default None)
     format: display format for this module (default '{layout}')
+    layouts: a list of layout that the user can select (default None)
```

Comment:
`(default [])` ? We do this for `thresholds` and all `list` config too. Just a heads up.

Reply:
yeah we need consistency in things.  We also need some guidance for module writers.

I think the best approach is to write the guidance and then apply it to modules.  If you want to make a start then that would be great.  Bear in mind that some of it will be easily agreed and some will take longer so we may have to create it in stages and add new parts incrementally.

Reply:
1) Abbrevs
    Do not use abbreviations?
    * e.g., temp --> temperature
    * e.g., prev --> previous

    Reason: Clean in the long run.
    Con: Long placeholders can be annoying in the format.
    Con: Does not sit well with other non-abbreviation words.
    Pro: Eliminate confusion about possible abbreviations by always avoiding
    the abbreviations. This becomes the traditional.

2) Defaults
    If string is not set, use None.
    * e.g., '' --> (default '')  REJECTED? --> (default None)
    * e.g., [] --> (default [])
    * e.g., {} --> (default {})
    * e.g.,  0 --> (default 0)

    Reason: Strings are more common than others so we don't reject others.
    Con: It's for the strings only. We could use (default '') which can be ugly.
    Pro: Users see (default None), they can assume it is a string without parsing
    the description, eg a list of layouts (default None) --> 'en, fr' ?

3) Config Names
    Avoid names that might not sit well with config prefix.
    Avoid names that might cause some confusion. Be i8ln-friendly.
    * e.g., button_select....
        We're already hovering.
        We're already selecting.
        We're already clicking.
        This changes nothing.

    We could use button_action to imply this will trigger an action. For others,
    it will not be an issue, e.g. button_next, button_mute, button_reset.

4) None

    We should not try to customize things if it's not part of default. This is
    more noticeable in update modules. If we default `UPD 0`, then `format_separator`
    is to be '' (default None) because if users want to add names (or update bars),
    they should decide if they want ', ' or ' ' or ' - ' rather than us.

    Update bars does not need separators. The names do. Lately, we have been favoring
    a space... which is a good default, but can be problematic undoing default configs
    so it may be better in the long run not to default anything.

    This is to avoid making users to do this... format_separator = ''.
    This may be noticeable in task_warrior, time_warrior, and some other modules I
    can't think of. I think window_manipulation was the first module.

5) Method Names

    The names should be clear and well defined. When you said `get_all_numbers`
    do you mean `get_all_numbers` or only some numbers? This may not always help,
    but can be useful. Think of it as writing comments as method names.

6) Examples
    All modules should have some examples. Monkey see, monkey do.
    We put in configs. Users copy configs. Simple facts.
    We put in excellent configs. Users uses excellent configs. Irrefutable facts.

7) Consistency
    This is the hard part. Symmetry is everything. Not only does symmetry makes the
    world go around, it also can makes most of the codes pretty. Coding is an art.
    Write nice codes. Get everything sorted out. Don't leave anything out.